unified_planning/plan.py

```diff
@@ -102,3 +114,101 @@ def __eq__(self, oth: object) -> bool:
+        '''Returns all the messages the planner gave about his activity.'''
+        return self._log_messages
+
+class IntermediateReport(Report):
+    '''Represents an intermediate report that a planner gives to the user through the clabback parameter of the Solver.solve() method.'''
+    def __init__(self, plan: Optional[Plan], planner_name: str = '', metrics: Dict[str, str] = {}, log_messages: List[str] = []):
+        Report.__init__(self, plan, planner_name, metrics, log_messages)
+    
+    def __repr__(self) -> str:
+        output = 'Intermediate Report:\n'
+        if self._planner_name != '':
+            output = f'planner: {self._planner_name}\n{output}'
+        output = f'{output}plan: {str(self._plan)}\n'
+        if self._metrics != {}:
+            metrics_str: str = ''
+            for mn, m in self._metrics.items():
+                metrics_str = f'{metrics_str}    {mn}: {m}\n'
+            output = f'{output}metrics: {metrics_str}'
+        if self._log_messages != []:
+            log_messages_str = "\n".join(self._log_messages)
+            output = f'{output}{log_messages_str}'
+        return output
+
+
+class FinalReport(Report):
+    '''Represents the final report that a planner gives to the user.'''
+    def __init__(self, status: int, plan: Optional[Plan], planner_name: str = '', metrics: Dict[str, str] = {}, log_messages: List[str] = []):
+        Report.__init__(self, plan, planner_name, metrics, log_messages)
+        self._status = status
+        # Checks that plan and status are consistent
+        if self._status in POSITIVE_OUTCOMES and self._plan is None:
+            raise #TODO: insert a proper exception. Or this should be an assert? Because it does not depend on USER but on SOLVEr
+        elif self._status in NEGATIVE_OUTCOMES and self._plan is not None:
+            raise #TODO: insert a proper exception
+    
+    def __repr__(self) -> str:
+        output = f'FinalReport:\n{status_to_str(self._status)}\n'
+        if self._planner_name != '':
+            output = f'planner: {self._planner_name}\n{output}'
+        output = f'{output}plan: {str(self._plan)}\n'
+        if self._metrics != {}:
+            metrics_str: str = ''
+            for mn, m in self._metrics.items():
+                metrics_str = f'{metrics_str}    {mn}: {m}\n'
+            output = f'{output}metrics: {metrics_str}'
+        if self._log_messages != []:
+            log_messages_str = "\n".join(self._log_messages)
+            output = f'{output}{log_messages_str}'
+        return output
+
+    def status(self) -> int:
+        '''Returns the status as an int.'''
+        return self._status
+    
+    def status_as_str(self) -> str:
+        '''Returns the status as a str.'''
+        return __STATUS_STR__[self._status]
+
+
+def status_to_str(id: int) -> str:
+    '''Returns a string representation of the given status.'''
+    return __STATUS_STR__[id]
+
+__STATUS_STR__ = {
+    SATISFIED: 'STATISFIED',
```

Comment:
Typo:
```suggestion
    SATISFIED: 'SATISFIED',
```

Check to use the same name here and in the proto file.

Reply:
Should we use SAT or SATISFIED? I personally prefer the long option (more clear), but it's not really something major.